proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+
+#### Matching Ambiguity
+
+As mentioned above, this design does introduce the possibility of
+matching ambiguity in certain cases, where there may be more than one
+variant of a resource that matches the dynamic parameters specified by
+the client.
+
+If an xDS transport protocol implementation does encounter multiple
+possible matching variants of a resource, its behavior is undefined.
+In the following sections, we evaluate the cases where that can occur
+and specify how each one will be addressed.
+
+##### Adding a New Key on the Server First
+
+Consider what would happen in the above transition scenario if we changed
+the server to have multiple variants of a resource differentiated by
+the new `version` key before all of the clients were upgraded to use
+that key.  For clients sending `{env=prod}`, there would be two possible
+matching variants of the resource, one for `version=v1` and another for
+`version=v2`, and there would be no way to determine which variant to
+use for that client.
+
+As stated above, we are optimizing for the case where new keys are added
+on clients first, since that is expected to be the common scenario.
+However, there may be cases where it is not feasible to have all clients
+start sending a new key before the server needs to start making use of
+that key.
+
+For example, let's say that this transition scenario is occurring in
+an environment where the xDS server is controlled by one team and the
+clients are controlled by various other teams, so it's not feasible to
+force all clients to start sending the new `version` key all at once.
+But there is one particular client team that is eager to start using
+the new `version` key to differentiate the configs of their clients,
+and they don't want to wait for all of the other client teams to start
+sending the new key.
+
+Consider what happens if the server simply adds a variant of the
+resource with the new key, while leaving the original resource variant
+in place:
+
+```textproto
+// Existing variant for older clients that are not yet sending the
+// version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+
+// New variant intended for clients sending the version key.
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+This will work fine for older clients that are not yet sending the
+`version` key, because their dynamic parameters will not match the new
+variant's constraints.  However, newer clients that are sending dynamic
+parameters `{env=prod, version=v1}` will run into ambiguity: those
+parameters can match either of the above variants of the resource.
+
+This situation will be avoided via a best practice that all authoritative
+xDS servers should have **all variants of a given resource specify
+constraints for the same set of keys**.
```

Comment:
Thinking out loud, couldn't this case also be handled by recommending that the server return the most specific match (the most keys that match)? This would not handle overlapping cases, but would handle this case well.

Reply:
As you say, that would work in only a subset of cases, so even if we did that, we'd need this best practice anyway.  And once we have this best practice, we don't need that other one.